XenonCode.hpp

```diff
@@ -2742,41 +2763,48 @@ const int VERSION_PATCH = 0;
 				assert(currentFunctionName == "");
+				ram_numericVariables_offset = ram_numericVariables;
+				ram_textVariables_offset = ram_textVariables;
+				ram_objectReferences_offset = ram_objectReferences;
+				ram_numericArrays_offset = ram_numericArrays;
+				ram_textArrays_offset = ram_textArrays;
+		
 				if (name != "system.timer" && name != "system.input" && !name.starts_with("entrypoint.") && functionRefs.contains(name)) {
 					throw CompileError("Function " + name + " is already defined");
 				}
 				currentFunctionName = name;
 				currentFunctionAddr = addr();
+				if (currentFunctionName == "system.timer") {
+					assert(currentTimerInterval);
+					timers.emplace_back(currentTimerInterval, currentFunctionAddr);
+					currentTimerInterval = 0;
+					userVars["system.timer"].clear();
+				} else if (currentFunctionName == "system.input") {
+					inputs[currentInputPort].addr = currentFunctionAddr;
+					currentInputPort = 0;
+					userVars["system.input"].clear();
+				} else if (currentFunctionName.starts_with("entrypoint.")) {
+					assert(entryPoints.size() > 0);
+					entryPoints.back().addr = currentFunctionAddr;
+					userVars[currentFunctionName].clear();
+				} else if (currentFunctionName != "") {
+					functionRefs.emplace(currentFunctionName, currentFunctionAddr);
+				}
 			};
 			auto closeCurrentFunction = [&](){
 				if (currentFunctionName != "") {
 					write(RETURN);
-					if (currentFunctionName == "system.timer") {
-						assert(currentTimerInterval);
-						timers.emplace_back(currentTimerInterval, currentFunctionAddr);
-						currentTimerInterval = 0;
-						userVars["system.timer"].clear();
-					} else if (currentFunctionName == "system.input") {
-						inputs[currentInputPort].addr = currentFunctionAddr;
-						currentInputPort = 0;
-						userVars["system.input"].clear();
-					} else if (currentFunctionName.starts_with("entrypoint.")) {
-						assert(entryPoints.size() > 0);
-						entryPoints.back().addr = currentFunctionAddr;
-						userVars[currentFunctionName].clear();
-					} else {
-						functionRefs.emplace(currentFunctionName, currentFunctionAddr);
-					}
 					currentFunctionName = "";
-					currentFunctionAddr = 0;
 				}
+				currentFunctionAddr = 0;
+				currentFunctionRecursive = false;
 				currentStackId = 0;
 			};
 			// If it's a user-declared function and it has a return type defined, returns the return var ref of that function, otherwise returns VOID
-			auto compileFunctionCall = [&](Word func, const std::vector<ByteCode>& args, bool getReturn, bool isTrailingFunction = false) -> ByteCode {
+			auto compileFunctionCall = [&](Word func, const std::vector<ByteCode>& args, bool getReturn, bool isTrailingFunction = false, bool recursive = false) -> ByteCode {
 				std::string funcName = func;
 				if (func == Word::Funcname) {
-					if (!functionRefs.contains(funcName)) {
+					if (!functionRefs.contains(funcName) || (funcName == currentFunctionName && !recursive)) {
```

Comment:
this allows out-of-order function calls, which removes the protection for anti-recursion calls even for non-recursive functions. I don't think this change is needed.
Also, given the `recursive` argument, it knows what function to call because that's the current scope and we can assume the `self` keyword was used since it's the only possible recursion.